src/deadlock-detection/lockorder.cpp

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2019 Greg Griffith
+// Copyright (c) 2019 The Bitcoin Unlimited developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "lockorder.h"
+#include "util.h"
+
+#ifdef DEBUG_LOCKORDER // this ifdef covers the rest of the file
+
+void CLockOrderTracker::potential_lock_order_issue_detected(const CLockLocation &thisLock,
+    const CLockLocation &otherLock,
+    const uint64_t &tid)
+{
+    LOGA("POTENTIAL LOCK ORDER ISSUE DETECTED\n");
+    LOGA("This occurred while trying to lock: %s after %s \n", thisLock.GetMutexName().c_str(),
+        otherLock.GetMutexName().c_str());
+    LOGA("Thread with id %" PRIu64
+         " attempted to lock %s on line %i in file %s after locking %s on line %i in file %s\n",
+        tid, thisLock.GetMutexName().c_str(), thisLock.GetLineNumber(), thisLock.GetFileName().c_str(),
+        otherLock.GetMutexName().c_str(), otherLock.GetLineNumber(), otherLock.GetFileName().c_str());
+    LOGA("We have previously locked these locks in the reverse order\n");
+    LOGA("\n\nOur historical lock orderings containing these two locks by thread %" PRIu64 " include: \n", tid);
+
+    std::pair<std::string, std::string> key = std::make_pair(otherLock.GetMutexName(), thisLock.GetMutexName());
+    auto iter = SeenLockLocations.find(key);
+    if (iter != SeenLockLocations.end())
+    {
+        for (auto &entry : iter->second)
+        {
+            if (std::get<2>(entry) == tid)
+            {
+                LOGA("this thread previously locked %s on %s after locking %s on %s\n", key.first.c_str(),
+                    std::get<0>(entry).c_str(), key.second.c_str(), std::get<1>(entry).c_str());
+            }
+        }
+        LOGA("\n\n");
+        LOGA("Our historical lock orderings containing these two locks by other threads include: \n");
+        for (auto &entry : iter->second)
+        {
+            const uint64_t storedTid = std::get<2>(entry);
+            if (storedTid != tid)
+            {
+                LOGA("Thread with id %" PRIu64 " previously locked %s on %s after locking %s on %s\n", storedTid,
+                    key.first.c_str(), std::get<0>(entry).c_str(), key.second.c_str(), std::get<1>(entry).c_str());
+            }
+        }
+        LOGA("\n\n");
+    }
+    throw std::logic_error("potential lock order issue detected");
+}
+
+bool CLockOrderTracker::CanCheckForConflicts(const std::string &lockname)
+{
+    return (SeenLockOrders.find(lockname) != SeenLockOrders.end());
```

Comment:
needs std::lock_guard<std::mutex> lock(lot_mutex); so other threads don't modify the map while find traverses it.

Reply:
fixed